tests/tracer/test_logging.py

```diff
@@ -0,0 +1,518 @@
+    assert b"warning log" in err
+    assert b"debug log" in err
+    assert out == b""
+
+    code = """
+import logging
+
+logging.basicConfig(format='%(message)s')
+ddtrace_logger = logging.getLogger('ddtrace')
+
+assert ddtrace_logger.getEffectiveLevel() == logging.DEBUG
+assert len(ddtrace_logger.handlers) == 0
+
+ddtrace_logger.warning('warning log')
+ddtrace_logger.debug('debug log')
+"""
+
+    out, err, status, pid = ddtrace_run_python_code_in_subprocess(code, env=env)
+    assert status == 0, err
+    assert re.search(LOG_PATTERN, str(err)) is None
+    assert "program executable" in str(err)  # comes from ddtrace-run debug logging
+    assert b"warning log" in err
+    assert b"debug log" in err
+    assert out == b""
+
+
+@pytest.mark.parametrize("dd_trace_log_file_level", ["CRITICAL", "ERROR", "WARNING", "INFO", "DEBUG", None])
+def test_debug_logs_can_go_to_file_backup_count(
+    dd_trace_log_file_level, run_python_code_in_subprocess, ddtrace_run_python_code_in_subprocess, tmpdir
+):
+    """
+    When DD_TRACE_DEBUG is true and DD_TRACE_LOG_FILE has been specified, debug logs are
+        written to a file at the expected backup count, based on the DD_TRACE_LOG_FILE_LEVEL setting.
+        Note: When running ddtrace-run, the ddtrace-run logs still emit to stderr.
+    """
+    env = os.environ.copy()
+    if dd_trace_log_file_level is not None:
+        env["DD_TRACE_LOG_FILE_LEVEL"] = dd_trace_log_file_level
+
+    log_file = tmpdir.strpath + "/testlog.log"
+    env["DD_TRACE_LOG_FILE"] = log_file
+    env["DD_TRACE_DEBUG"] = "true"
+    env["DD_TRACE_FILE_SIZE_BYTES"] = "10"
+    code = """
+import logging
+import os
+import ddtrace
+
+ddtrace_logger = logging.getLogger('ddtrace')
+assert ddtrace_logger.getEffectiveLevel() == logging.DEBUG
+assert len(ddtrace_logger.handlers) == 1
+assert isinstance(ddtrace_logger.handlers[0], logging.handlers.RotatingFileHandler)
+assert ddtrace_logger.handlers[0].maxBytes == 10
+assert ddtrace_logger.handlers[0].backupCount == 1
+if os.environ.get("DD_TRACE_LOG_FILE_LEVEL") is not None:
+    ddtrace_logger.handlers[0].level == getattr(logging, os.environ.get("DD_TRACE_LOG_FILE_LEVEL"))
+
+ddtrace_logger = logging.getLogger('ddtrace')
+
+for attempt in range(100):
+    ddtrace_logger.debug('ddtrace multiple debug log')
+    ddtrace_logger.critical('ddtrace multiple debug log')
+"""
+
+    out, err, status, pid = run_python_code_in_subprocess(code, env=env)
+    assert status == 0, err
+
+    if PY2:
+        assert 'No handlers could be found for logger "ddtrace' in err
+    else:
+        assert err == b""
+
+    assert out == b""
+
+    testfiles = os.listdir(tmpdir.strpath)
+    log_files = [filename for filename in testfiles if "testlog.log" in filename]
+    log_files.sort()
+    assert log_files == ["testlog.log", "testlog.log.1"]
+
+    with open(log_file) as file:
+        content = file.read()
+        assert len(content) > 0
+        assert re.search(LOG_PATTERN, content) is not None
```

Comment:
This is repeated a number of times, I wonder if we can extract into a helper function?

Reply:
I tried to extract out the repeated portions out, let me know if this looks cleaner to go through now!

Reply:
Marking as resolved.